opentelemetry/src/global/trace.rs

```diff
@@ -285,9 +268,275 @@ where
+}
+
+#[cfg(test)]
+// Note that all tests here should be marked as ignore so that it won't be picked up by default
+// We need to run those tests one by one as the GlobalTracerProvider is a shared object between threads
+// Use cargo test -- --ignored --run_threads=1 to run those tests.
+mod tests {
+    use super::*;
+    use crate::trace::NoopTracer;
+    #[cfg(any(feature = "tokio-support", feature = "tokio-rt-current-thread"))]
+    use crate::trace::Tracer;
+    use std::fmt::Debug;
+    #[cfg(any(feature = "tokio-support", feature = "tokio-rt-current-thread"))]
+    use std::io::Write;
+    #[cfg(any(feature = "tokio-support", feature = "tokio-rt-current-thread"))]
+    use std::sync::Mutex;
+    use std::thread;
+    use std::thread::sleep;
+    use std::time::Duration;
+
+    macro_rules! cfg_tokio {
+        ($($item:item)*) => {
+            $(
+                #[cfg(any(feature = "tokio-support", feature = "tokio-rt-current-thread"))]
+                $item
+            )*
+        }
+    }
+
+    cfg_tokio! {
+        #[derive(Debug)]
+        struct AssertWriter {
+            buf: Arc<Mutex<Vec<u8>>>
+        }
+
+        impl AssertWriter {
+            fn new() -> AssertWriter {
+                AssertWriter {
+                    buf: Arc::new(Mutex::new(Vec::new())),
+                }
+            }
+
+            fn len(&self) -> usize {
+                self.buf.lock().expect("cannot acquire the lock of assert writer").len()
+            }
+        }
+
+        impl Write for AssertWriter {
+            fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+                let mut buffer = self.buf.lock().expect("cannot acquire the lock of assert writer");
+                buffer.write(buf)
+            }
+
+            fn flush(&mut self) -> std::io::Result<()> {
+                let mut buffer = self.buf.lock().expect("cannot acquire the lock of assert writer");
+                buffer.flush()
+            }
+        }
+
+        impl Clone for AssertWriter {
+            fn clone(&self) -> Self {
+                AssertWriter {
+                    buf: self.buf.clone()
+                }
+            }
+        }
+    }
+
+    #[derive(Debug)]
+    struct TestTracerProvider {
+        debug_msg: &'static str,
+    }
+
+    impl Default for TestTracerProvider {
+        fn default() -> Self {
+            TestTracerProvider { debug_msg: "" }
+        }
+    }
+
+    impl TestTracerProvider {
+        fn new(debug_msg: &'static str) -> Self {
+            TestTracerProvider { debug_msg }
+        }
+    }
+
+    impl TracerProvider for TestTracerProvider {
+        type Tracer = NoopTracer;
+
+        fn get_tracer(&self, _name: &'static str, _version: Option<&'static str>) -> Self::Tracer {
+            NoopTracer::default()
+        }
+    }
+
+    #[test]
+    #[ignore]
+    fn test_set_tracer_provider() {
+        let _guard = set_tracer_provider(TestTracerProvider::new("global one"));
```

Comment:
There are a few more instances of the `_guard` variable here in in the following lines/tests. Not sure if you want to replace them with `_previous` or just `_`.

Reply:
Oops, I think I forget to clean the remaining `_guard`. Should be fixed now.